### ISSUE TITLE
[feature request] fast interpolation in uniformly-spaced data

### ISSUE THREAD

**Issue by Phillip-M-Feldman** [scipy/scipy#10180]
When working with uniformly-spaced data, it is possible to skip the initial search operation that an interpolation algorithm would ordinarily have to perform.  This tends to speed things up significantly.  It would be nice to have some specialized interpolators in SciPy that do linear interpolation and spline interpolation on uniformly-spaced data.  The following sample code does linear interpolation without search:

```
   # Interpolate linearly, taking advantage of the fact that
   # the values in `f` are uniformly spaced.

   quo, rem= np.divmod(x, step)
   fraction= rem / step

   # Although the quotient values returned by `numpy.divmod` are integers, the
   # array dtype is 'f4'.  ==> We must force the dtype to 'i4' so that the
   # quotient values can be used for indexing:
   quo= quo.astype('i4')

   result= fraction * r[quo+1] + (1-fraction) * r[quo]
```

**Comment by ev-br** [comment]
This feels best suited for specialized user code IMO.

**Comment by Phillip-M-Feldman** [comment]
Over the last 40 years, most of the data with which I've worked has been
uniformly spaced in the independent variables.  So, I'm inclined to think
that this case is not that specialized.

**Comment by larsoner** [comment]
It sounds like you're describing a regular grid in 1D. Can what you want (at least for linear mode) be done with [RegularGridInterpolator](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.RegularGridInterpolator.html)? If so, is it close enough in performance to your simple code?

If you need splines, maybe it's easy to extend it beyond `linear` and `nearest` modes to some spline ones.

**Comment by WarrenWeckesser** [comment]
@larsoner, `RegularGridInterpolator` doesn't assume that the grid spacing is uniform.  From the docstring: "The data must be defined on a regular grid; the grid spacing however may be
uneven."

**Comment by Phillip-M-Feldman** [comment]
"Regular grid" and "uneven spacing" seems like a contradiction.

**Comment by larsoner** [comment]
> > RegularGridInterpolator doesn't assume that the grid spacing is uniform. From the docstring: "The data must be defined on a regular grid; the grid spacing however may be uneven."

> "Regular grid" and "uneven spacing" seems like a contradiction.

My initial reading of this (admittedly confusing) docstring was that the spacing *across dimensions* could vary (e.g., 1. along the first dim, and 2. along the second), but that the spacing *within a given dimension* must be equal from point to point. (Think non-equal-aspect-ratio pixels.) But @WarrenWeckesser playing with examples I think you're right that all it assumes is that the spacing essentially broadcasts across dimensions. So the grid is "regular" in this broadcasting sense, not in a uniform-within-dimension-spacing sense (or what I would take to be the strictest interpretation, uniform-across-dimensions spacing sense).

I agree it would be useful to add something like `assume_uniform` or `check_uniform` to see if it's possible to skip the search using the fixed per-dimension `delta` of the `points`.

**Comment by Phillip-M-Feldman** [comment]
This makes sense.  Thanks!

**Comment by huizew** [comment]
> I agree it would be useful to add something like `assume_uniform` or `check_uniform` to see if it's possible to skip the search using the fixed per-dimension `delta` of the `points`.

I can try to implement this if no one is working on it yet.

**Comment by rgommers** [comment]
@huizew please go for it!

**Comment by MatthewFlamm** [comment]
I've gone down this road in the past.  I implemented a version similar to Phillip's method, but it was slower than using numpy.searchsorted!  It was slower in my testing even on vectors of lengths over 1e6.  I wonder if numpy.searchsorted has some internal uniformity check for speedup already.  Note that the current implementation searches over each dimension independently, so a vector test is suitable here.

I suggest any addition have a very comprehensive timing comparison to make sure that it is worth it.

**Comment by MatthewFlamm** [comment]
Time to eat my words :smile: .  My previous foray must have been for single point interpolations rather than vectorized.  I still suggest comprehensive timing to make sure that it is indeed faster.

```
import numpy as np 
x, dx = np.linspace(0, 1, 10000000, retstep=True) 
x_r = np.random.rand(10000) 
 
def searchuniform(xp, dx): 
    i, _ = np.divmod(xp, dx) 
    return i.astype('int64') 
%timeit np.searchsorted(x, x_r) 
%timeit searchuniform(x_r, dx)

3.45 ms ± 10.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
148 µs ± 509 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Comment by huizew** [comment]
@MatthewFlamm Thanks! I agree that comprehensive timing is needed.

In terms of time complexity analysis, ``searchsorted`` should be O(logn) while search in uniform grid should be O(1), for 1-D case. A "uniformity check" is O(n) and I believe it should be run only once, during the interpolation class object initialization.

**Comment by Phillip-M-Feldman** [comment]
In all of my use cases, I define the interpolating function and then apply
it exactly once, so it would be helpful if there were a way to pass a flag
that says, "The caller promises that the independent variable is uniformly
spaced."

**Comment by ev-br** [comment]
For 1D splines I think a reasonable way to go would be to add something like `CardinalBSpline` class to mirror the general BSpline. Instead of the knot array, this object would have e.g. the grid step size and implement the specialized version of the `_find_interval` (via `np.divmod` or some such). Furthermore, low-order spline basis elements can be hardcoded to be more efficient than the general de Boor algorithm which `BSpline` uses.

**Comment by ev-br** [comment]
Cross-ref https://github.com/scipy/scipy/pull/17317, which documents a long-available if not-widely-known wrapper over ndimage.map_coordinates.

**Comment by ev-br** [comment]
Now that https://github.com/scipy/scipy/pull/17317 is merged and the recipe is available at https://scipy.github.io/devdocs/tutorial/interpolate/ND_regular_grid.html#uniformly-spaced-data, enhancements for data with equal spacing should start from investigating wrappers similar to the one listed in this doc like. Can a listcomp can be changed to something more performant? Is it difficult to adapt the simplistic example to real-world usage? Are there further performance enhancements which are difficult to do in userland?
In principle, we are open to adding things to the scipy.interpolate namespace--- it would need a compelling case though.
